alainmateso/iFind
# Ft user signup 168532824

#### What does this PR do?

This PR implements the following endpoints:

- User sign up
- User sign in 

In order to achieve this, it was necessary to create the following helper files:

```javascript
.../ helpers 
      hashHelper.js
      responseHelper.js
      stringsHelper.js
      tokenHelper.js
```

#### Description of Task to be completed?

###### A user must be able to sign up to the lost and found platform.

This is achieved when a user access the following endpoint

` api/v1/auth/signup ` using the `   POST  ` method.

In order to sign up a user must have:

- first_name | String - required
- last_name | String - required
- email | String - required
- phonenumber | String - required
- password | String - required

When the user signs up successfully they are returned with a response of:

```javascript
{
   status: 201,
   message: 'Successfully created user', 
   data: {
             token: tokenString
   }     
}
```

If the signup is unsuccessful due to a problem with the request, the user is returned with a response of:

```javascript
{
   status: 400,
   message: 'Unable to sign up user', 
   data: {
             error: errorMessageString
   }     
}
```
If the signup is unsuccessful due to the email already existing in the system, the user is returned with a response of:

```javascript
{
   status: 409,
   message: 'User with this email already exists', 
}
```


###### A user must be able to sign in to the lost and found platform.

This is achieved when a user access the following endpoint

` api/v1/auth/signin ` using the `   POST  ` method.

In order to sign in a user must have:

- email | String - required
- password | String - required

When the user signs in successfully they are returned with a response of:

```javascript
{
   status: 200,
   message: 'Successfully logged in user', 
   data: {
             token: tokenString
   }     
}
```

If the signin is unsuccessful due to a problem with the request data, the user is returned with a response of:

```javascript
{
   status: 400,
   message: 'Unable to sign up user', 
   data: {
             error: errorMessageString
   }     
}
```
If the signin is unsuccessful due to an incorrect email, password combination, the user is returned with:

```javascript
{
   status: 404,
   message: 'Unable to sign in user, check info', 
}
```

#### How should this be manually tested?

1. Clone the Github repository
2. Run npm install to install development dependencies.
3. Create a ` .env ` file in root directory of the project and fill in the following:
     - DB_NAME = NAME_OF_YOUR_DATABASE
     - TEST_DB = NAME_OF_YOUR_TEST_DATABASE
     - DB_USER = NAME_OF_THE_USER_FOR_YOUR_DATABASE
     - DB_PASS = PASSWORD_OF_YOUR_DATABASE_USER
     - DB_PORT = DATABASE_PORT || 5432 (Default Postgres Port)
     - DB_HOST= HOST_RUNNING_YOUR_DB || 127.0.0.1 (Default host)

SECRET_KEY = SECRET_KEY_FOR_PASSWORD_HASHING

4. Run the ` sequelize db:migrate ` command migrate db tables.

5. Test the endpoints and ensure correctness and program sanity. 

#### Any background context you want to provide?

Previously, no helper files were used to return responses and strings. These files were created to ensure the modularity of the code. 

##### The following files have been created to be used across the application
1. Strings Helper ('helpers/stringsHelper.js')

This file returns an object of sub-objects and strings. 

* For Example *
```javascript
const strings = {
  users: {
    errorMessages: {
      BAD_SIGNUP_REQUEST: 'Unable to sign up user',
    },
}
```

This string can be accessed in code by simply importing the file and accessing it like so:

```javascript
import strings from 'stringsHelper.js'

return (res.status(500).json({
    message: strings.user.BAD_SIGNUP_REQUEST
```

2. Response Helper ('helpers/responseHelper.js'`)

This file is a response utility aimed at simplifying responses to the user from any part of the application. 

The helper is structured as follows:

```javascript

export default (res, status, message, data) => res.status(status).json({
  status,
  message,
  data,
});

```
The function takes in:

- res: A response to return responses to the user
- status: A status To be displayed in the body and as a response status
- message: A message to be displayed to the user
- data: Any data the user must-see. If not applicable, this param can be left blank

#### What are the relevant pivotal tracker stories?

##### #168532824 
##### #168532778

#### Screenshots (if appropriate)

<img width="1369" alt="Screen Shot 2019-09-18 at 08 53 20" src="https://user-images.githubusercontent.com/13886438/65127355-97e49400-d9f7-11e9-873a-38b56c5b8dd5.png">
<img width="1377" alt="Screen Shot 2019-09-18 at 08 53 59" src="https://user-images.githubusercontent.com/13886438/65127358-987d2a80-d9f7-11e9-8213-409b810d0f66.png">
<img width="1368" alt="Screen Shot 2019-09-18 at 08 54 11" src="https://user-images.githubusercontent.com/13886438/65127366-99ae5780-d9f7-11e9-9f74-3de1a9b3da90.png">
<img width="1367" alt="Screen Shot 2019-09-18 at 08 54 19" src="https://user-images.githubusercontent.com/13886438/65127369-99ae5780-d9f7-11e9-8c67-aa46be3e3c10.png">
<img width="1371" alt="Screen Shot 2019-09-18 at 08 54 28" src="https://user-images.githubusercontent.com/13886438/65127373-99ae5780-d9f7-11e9-9db4-f82dddd63fdf.png">
<img width="1370" alt="Screen Shot 2019-09-18 at 08 54 41" src="https://user-images.githubusercontent.com/13886438/65127376-9a46ee00-d9f7-11e9-91f7-41ce5eb59fb1.png">

#### Questions:

N / A
